First draft of an update to the KPO guide (#388)
astronomer/airflow-guides: guides/kubepod-operator.md

The `kubernetes_conn_id` is the preferred way of setting up your Kubernetes connection. `in_cluster`, `cluster_context` and the path to the `config_file` can be set within your connection and overwritten by using the parameters in the KPO. Setting these parameters at the level of `airflow.cfg` has been deprecated.

### Example: Using the KubernetesPodOperator to run a script in Haskell

A frequent use case for the KubernetesPodOperator is to run scripts for tasks better described in a language other than Python. For this purpose a custom Docker image has to be build and either run from either a public or private [DockerHub repository](https://docs.docker.com/docker-hub/repos/).

> **Note**: Astro provides documentation on [how to run images from private repositories](https://docs.astronomer.io/astro/kubernetesPodoperator).

The code below shows a Haskell script which takes the environment variable `NAME_TO_GREET` and prints a message containing it to the console:

```haskell
import System.Environment

main = do
        name <- getEnv "NAME_TO_GREET"
        putStrLn ("Hello, " ++ name)
```

> The Dockerfile uses `cabal` as a system for building and packaging Haskell programs, for this example the `haskell_example.cabal` file built automatically when running `cabal init` in a directory named `haskell_example` is sufficient.

Reviewer: ```suggestion
The Dockerfile creates the necessary environment to run the script and then executes it with a `CMD` command.
```
I know the context you had is relevant for the actual content of the example, but for the purposes of what we're trying to show I think we can keep it more general.

```Dockerfile
FROM haskell
WORKDIR /opt/hello_name
RUN cabal update
COPY ./haskell_example.cabal /opt/hello_name/haskell_example.cabal
RUN cabal build --only-dependencies -j4
COPY . /opt/hello_name
RUN cabal install
CMD ["haskell_example"]
```

> After making the Docker image available it is possible to launch it from the KPO via the `image` argument. The example DAG below showcases a variety of arguments of the KubernetesPodOperator including how to pass the environment variable `NAME_TO_GREET` which is used from within the Haskell code.

Reviewer: ```suggestion
After making the Docker image available it can be run from the KPO via the `image` argument. The example DAG below showcases a variety of arguments of the KubernetesPodOperator, including how to pass the environment variable `NAME_TO_GREET` which is used from within the Haskell code.
```

        labels={'app':'backend', 'env':'dev'},
        # reattach to worker instead of creating a new Pod on worker failure
        reattach_on_restart=True,
        # delete Pod after the task is finished
        is_delete_operator_pod=True,
        # get log stdout of the container as task logs
        get_logs=True,
        # log events in case of Pod failure
        log_events_on_failure=True,
        # pass your name as an environment var
        env_vars={"NAME_TO_GREET": f"{name}"}
        )
```

## KubernetesPodOperator best practices

SECTION INCOMPLETE

Suggestions from issues:  

- CI/CD guidelines (such as you need to be pushing updates to the KPO image)
- make sure each KPO image update has a unique tag
Some advice on where to store the images such that they can be retrieved on Astro
- how to best set env variables

## Example: Using the KubernetesPodOperator with XComs

[XComs](https://airflow.apache.org/docs/apache-airflow/stable/concepts/xcoms.html) is a commonly used way to pass small amounts of data between tasks. It is possible to use this feature with the KPO both for it to receive values stored in XCom and push values to XComs itself.

> The DAG below shows a straightforward ETL pipeline with an `extract_data` task simulating running a query on a database and returning one value that by being returned from the Python function is automatically pushed to XComs thanks to the [TaskFlow API](https://airflow.apache.org/docs/apache-airflow/stable/tutorial_taskflow_api.html#tutorial-on-the-taskflow-api).  

Reviewer: ```suggestion
The DAG below shows a straightforward ETL pipeline with an `extract_data` task that runs a query on a database and returns a value; the return value is automatically pushed to XComs thanks to the [TaskFlow API](https://airflow.apache.org/docs/apache-airflow/stable/tutorial_taskflow_api.html#tutorial-on-the-taskflow-api).  
```

> The `transform` task is a KubernetesPodOperator which launches an image having been built with the following Dockerfile:

Reviewer: ```suggestion
The `transform` task is a KubernetesPodOperator which requires the XCom data pushed from the upstream task before it, and launches an image having been built with the following Dockerfile:
```

        # Pod configuration
        name='my_Pod', # naming the Pod
        get_logs=True, # log stdout of the container as task logs
        log_events_on_failure=True, #log events in case of Pod failure
        env_vars={"DATA_POINT": "{{ ti.xcom_pull(task_ids='extract_data', key='return_value') }}"},
        do_xcom_push=True #push the contents from xcom.json to xcoms
        )

    @task
    def load_data(**context):
        transformed_data_point = context['ti'].xcom_pull(task_ids='extract_data', key='return_value')
        print(transformed_data_point)

    extract_data() >> transform >> load_data()
```

## Example: Spinning up a Pod in EKS from Airflow

If some of your tasks require specific resources like a GPU you may want to run them in a different cluster than your Airflow Instance. In setups where both clusters are used by the same AWS or GCP account this can be managed with roles and permissions. There is also the possibility to use a CI account and enable [cross-account access to AWS EKS cluster resources](https://aws.amazon.com/blogs/containers/enabling-cross-account-access-to-amazon-eks-cluster-resources/).  

> However you may want to use a specific AWS or GCP account that you keep separate from your Airflow environment. Or you are currently running Airflow on a local setup not using Kubernetes but still want to run some tasks in a Kubernetes Pod on a remote cluster, these are possible use cases for this example.

Reviewer: ```suggestion
However, you may want to use a specific AWS or GCP account that you keep separate from your Airflow environment. Or you are currently running Airflow on a local setup not using Kubernetes, but still want to run some tasks in a Kubernetes pod on a remote cluster.
```

> The example below shows the steps to take to set up an EKS cluster on AWS and run a Pod on it from an Airflow instance if cross-account access is not feasible. After setup of the EKS cluster the `KubeConfig` as well as the `AWS Profile` have to be provided to Airflow to make the connection to the remote cluster.  

Reviewer: ```suggestion
The example below shows the steps to take to set up an EKS cluster on AWS and run a pod on it from an Airflow instance where cross-account access is not feasible. After setup of the EKS cluster, the `KubeConfig` as well as the `AWS Profile` have to be provided to Airflow to make the connection to the remote cluster. Note that the same general steps are applicable to other Kubernetes services (e.g. GKE).  
```

> > **Note**: If you are using [Google Kubernetes Engine](https://cloud.google.com/kubernetes-engine/) consider using [GKEStartPodOperator](https://registry.astronomer.io/providers/google/modules/gkestartPodoperator).

Reviewer: I'm not super familiar with these operators and how they work, and I don't think they're very common, so I'm wondering if we should not recommend them and remove these notes. what do you think?

Author: Hmm... I am unsure what is best. I've not used the GKEStartPodOperator but tried to get the EksPodOperator to run and it keeps not recognizing the name of my cluster; as far as I can tell both are just supposed to make the connection easier but not adding any functionality.
 
I think from a user perspective the note is not really necessary because most will use the KPO unless the specifically are looking for a platform-specific Operator in the registry. I mostly added the note for internal consistency because of the "use the more specific operator if it exists"-rule and because Jeff pointed out that I could use the EksPodOperator. 😅 

Maybe a compromise between recommending and not mentioning would be to say something along the lines of "these two more specific operators exist if you want to check them out here are the links"?

Reviewer: I agree with your idea - fine to point out they exist, but we won't explicitly recommend them.

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "AWS": "arn:aws:iam::<aws account id>:<your user>",
        "Service": [
              "ec2.amazonaws.com",
              "eks.amazonaws.com"
        ]
      },
      "Action": "sts:AssumeRole"
    }
  ]
}
```

Add the new role to your local AWS credentials file which by default can be found at `~/.aws/config`.

```text
[default]
region = <your region>

[profile <name of the new role>]
role_arn = <EKS role arn>
source_profile = <your user>
```

Make sure your default credentials in `~/.aws/credentials` are using a valid and active key for your username (keys can be generated at **IAM** -> **Users** -> your user -> **Security Credentials**).

Make a copy of `~/.aws` available to your Airflow environment (for Astro users: copy it into the `include` folder).

Lastly, create a new EKS cluster (**EKS** -> **Clusters** -> **Add cluster**) and assign the the newly created role to it. The cluster might take a few minutes to become active.

> ### Step 2 Retrieve the KubeConfig file from the EKS cluster

Reviewer: ```suggestion
### Step 2: Retrieve the KubeConfig file from the EKS cluster
```

```bash
aws eks --region region update-kubeconfig --name cluster_name
```

The command above will copy information relating to the new cluster into your existing `KubeConfig` file at `~/.kube/config`. Select the information pertaining to your EKS cluster and make it available to your Airflow environment (for Astro users: copy it into the `include` folder). It is likely that you will have to add the last 5 lines to get to the following structure:

```text
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: <your certificate>
    server: <your AWS server address>
  name: <arn of your cluster>
contexts:
- context:
    cluster: <arn of your cluster>
    user: <arn of your cluster>
  name: <arn of your cluster>
current-context: <arn of your cluster>
kind: Config
preferences: {}
users:
- name: <arn of your cluster>
  user:
    exec:
      apiVersion: client.authentication.k8s.io/v1alpha1
      args:
      - --region
      - <your cluster's AWS region>
      - eks
      - get-token
      - --cluster-name
      - <name of your cluster>
      - --profile
      - default
      command: aws
      interactiveMode: IfAvailable
      provideClusterInfo: false
```

> The `KubeConfig` file is how the KPO will connect to a Kubernetes cluster by running the awscli with the default profile, using the credentials you provided in step 1.

Reviewer: ```suggestion
The `KubeConfig` file is how the KPO will connect to a Kubernetes cluster by running the `awscli` with the default profile, using the credentials you provided in Step 1.
```

> ### Step 3 Add a namespace and service account to the EKS Cluster

Reviewer: ```suggestion
### Step 3: Add a namespace and service account to the EKS Cluster
```

> It is best practice to use a new namespace for all Pods sent to the EKS cluster from the Airflow environment and a separate service account.  

Reviewer: ```suggestion
It is best practice to use a new namespace and separate service account for all pods sent to the EKS cluster from the Airflow environment.  
```

```bash
# create a new namespace on the EKS cluster
kubectl create namespace <your namespace name>

# if you are using a service account other than default in the Kubernetes
# environment your Airflow instance is running it you will have to add a
# service account of the same name to your remote cluster
kubectl create serviceaccount <service account name> -n <your namespace name>
```

> ### Step 4 Adjust Airflow configuration files

Reviewer: ```suggestion
### Step 4: Adjust the Airflow configuration files
```

> This step will differ depending on the Airflow setup. When running Airflow locally it is just necessary to make sure `awscli`, `apache-airflow-providers-cncf-kubernetes`,

Reviewer: ```suggestion
This step will differ depending on the Airflow setup. When running Airflow locally, it is necessary to make sure `awscli`, `apache-airflow-providers-cncf-kubernetes`,
```

and `apache-airflow-providers-amazon` are installed correctly on the local machine.

> For dockerized settings add the following line to your Dockerfile to copy your aws credentials from `/include` (or any other location) into the container:

Reviewer: ```suggestion
If you are running Airflow with Docker, add the following line to your Dockerfile to copy your aws credentials from `/include` (or any other location) into the container:
```
Is this edit correct?

```dockerfile
COPY --chown=astro:astro include/.aws /home/astro/.aws
```

To be able to correctly authenticate yourself to the remote cluster it is also important to add the AWS command line tool (`awscli`) to your `packages.txt`, and the necessary provider packages to `requirements.txt` (these will differ depending on what cloud you are connecting to):

```text
awscli
```

```text
apache-airflow-providers-cncf-kubernetes
apache-airflow-providers-amazon
```

> ### Step 5 Add AWS connection ID

Reviewer: ```suggestion
### Step 5: Add AWS connection ID
```

> In the Airflow UI navigate to **Admin** -> **Connections** to set up the connection to the AWS account running the EKS. Chose a unique connection ID and use your `aws_access_key_id` as login and your `aws_secret_access_key` as password.

Reviewer: ```suggestion
In the Airflow UI navigate to **Admin** -> **Connections** to set up the connection to the AWS account running the target EKS cluster. Chose a unique connection ID and use your `aws_access_key_id` as login and your `aws_secret_access_key` as password.
```

![Adding AWS connection](https://assets2.astronomer.io/main/guides/your-guide-folder/aws_connection.png)

> ### Step 6 Add the DAG interacting with EKS  

Reviewer: ```suggestion
### Step 6: Create the DAG with the KPO 
```

If dynamic behavior of the EKS cluster is desired, e.g. Nodes get created with specifications for a task and deleted after the task has run, it is necessary to use several classes from the Amazon provider package as shown in the example DAG below. If your remote Kubernetes cluster has Nodes already available you will only need the KubernetesPodOperator itself (task number 3 in the example).

> The example DAG contains 5 consecutive tasks. The first one creates a Nodegroup according to the users' specifications and afterwards a sensor checks that the cluster is running correctly. The third tasks is the actual KubernetesPodOperator running any valid Docker image in a Pod on the newly created Nodegroup on the remote cluster. The example below uses the standard `Ubuntu` image to print "hello" to the console using a `bash` command. Lastly the Nodegroup gets deleted and the deletion gets verified.

Reviewer: Can we bullet or number list the descriptions of the 5 tasks? I try to follow Jake's rule of listing any time there are more than 3 items :)

        is_delete_operator_Pod=False,
        in_cluster=False,
        config_file='/usr/local/airflow/include/config',
        startup_timeout_seconds=240,
        # this line is specific to users of Astro and not needed in OSS Airflow
        labels={"kubernetes_pod_operator": "external-cluster"}
    )

    # task 4 deleting the nodegroup
    delete_gpu_nodegroup=EKSDeleteNodegroupOperator(
        task_id='delete_gpu_nodegroup',
        cluster_name='<your cluster name>',  
        nodegroup_name='gpu-nodes',
        aws_conn_id='<your aws conn id>',
        region='<your region>'
    )

    # task 5 checking that the nodegroup was deleted successfully
    check_nodegroup_termination=EKSNodegroupStateSensor(
        task_id='check_nodegroup_termination',
        cluster_name='<your cluster name>',
        nodegroup_name='gpu-nodes',
        aws_conn_id='<your aws conn id>',
        region='<your region>',
        mode='reschedule',
        timeout=60 * 30,
        target_state=NodegroupStates.NONEXISTENT
    )

    # setting the dependencies
    create_gpu_nodegroup >> check_nodegroup_status >> run_on_EKS
    run_on_EKS >> delete_gpu_nodegroup >> check_nodegroup_termination

> ```

Reviewer: Yes exactly :)

> **Note**: On Astro the infrastructure needed to run KPO with CeleryExecutor is pre-built into every Cluster. Astronomer provides [comprehensive documentation on using KPO on Astro](https://docs.astronomer.io/astro/kubernetespodoperator).  

## When to use KPO

The KubernetesPodOperator runs any Docker image provided to it, whether they are pulled from [DockerHub](https://hub.docker.com/) or from private repositories. Frequent use cases are:

- Running a task better described in a language other than Python.
- Having full control over how much compute resources and memory a single task can use.
- Executing tasks in an separate environment with individual packages and dependencies.

Sometimes you may want to run pods on different clusters, for example if only some need GPU resources while others do not. This is also possible with the KPO and a step-by-step example is provided at the end of this guide.

> ## How to configure KPO

Reviewer: I don't think you have too many parameters described here, but I do think we should intro this section with a bit more context on what the KPO is doing - e.g. the work the task actually does is run the docker image you provide, and all other parameters are specific to the pod configuration. And that through the parameters you have access to everything available in the k8s API and then link to that.

Reviewer: Also, I think it would be helpful to present some of these parameters in the context of a simple example. Maybe running a 'hello world' script in some other language - then talk about what would be in the image, how you would set up the pod, etc.

- `kubernetes_conn_id`: uses a [connection](https://www.astronomer.io/guides/connections) stored in the [Airflow metadata database](https://www.astronomer.io/guides/airflow-database), which can be configured in the Airflow UI under **Admin** -> **Connections**.
- `in_cluster`: by default set to `True`, which means the new pod will be spun up in the same cluster, you are running your Airflow instance in.
- `config_file`: The path to the Kubernetes config file. If not specified, default value is `~/.kube/config`.
- `cluster_context`: is used to specify a context that points to a Kubernetes cluster when `in_cluster` is set to `False`. This parameter is used to select a specific cluster if several are defined within your config file.

The `kubernetes_conn_id` is the preferred way of setting up your Kubernetes connection. `in_cluster`, `cluster_context` and the path to the `config_file` can be set within your connection and overwritten by using the parameters in the KPO. Setting these parameters at the level of `airflow.cfg` has been deprecated.

> ## KPO vs KubernetesExecutor

Reviewer: Maybe this section should go as a subsection under When to Use KPO? I think that would flow well if we expand on the use cases a bit

The [KubernetesExecutor](https://airflow.apache.org/docs/apache-airflow/stable/executor/kubernetes.html) and the KubernetesPodOperator both dynamically launch and terminate Pods to run Airflow tasks in. As the names suggest, KubernetesExecutor is an executor, which means it affects how all tasks in an Airflow instance are executed while the KubernetesPodOperator is an operator defining that a single task is to be launched in a Kubernetes pod with the given configuration, which does not affect any other tasks in the Airflow instance.

Compared to the KPO the KubernetesExecutor:

- is set at the configuration level of the Airflow instance, which means _all_ tasks will each be run in their own Kubernetes pod.
- has less abstraction over pod configuration. All configurations have to be passed in as a dictionary via the argument `executor_config`.
- creates a new field in the view of individual task instances in the Airflow UI `K8s Pod Spec` which shows the specifications of the pod that was run.
- can only use Docker images that have Airflow installed, otherwise they will not be able to run the task. This is not the case with the KPO, which can run any valid Docker image.

> ## KPO local development

Reviewer: Now that I'm reading this, for general flow I think it would make sense to have this as a subsection under Requirements to Use KPO. So at a high level we would have infra/environment requirements -> when/how to use the operator -> best practices and specific examples

Reviewer: Another comment on this section - for the instructions, we should probably either make them more detailed or keep it very high level and not have steps. I'm kind of leaning towards the latter, and keeping this section pretty short and maybe linking out to other resources (so we can focus the guide more on the operator itself and use cases). But would love your opinion here having just learned it and worked with it.

Author: Yes I agree, a more detailed explanation would feel out of place here and distract from the KPO itself. 

I would have been lost without Marc's blogpost in setting this up (the official docs only provide instructions on how to install the helm chart but not the full setup) and just found a webinar on the Astronomer channel containing the same content. I think linking to that works? :)

Reviewer: Ah I had forgotten we did a webinar on that - great idea to link out to that one :)

heroImagePath: null
tags: ["Kubernetes", "Operators"]
---

## Overview

> The [`KubernetesPodOperator`](https://registry.astronomer.io/providers/kubernetes/modules/kubernetesPodoperator) (KPO) is an operator from the Kubernetes provider package designed to provide a straightforward way to execute a task in a Kubernetes Pod from your Airflow environment. In a nutshell, the KPO is an abstraction over a call to the Kubernetes API to launch a Pod.

Author: Hmm this might be me but I would still read this the wrong way around as "write your task in airflow and it will be put into a docker image to be run in a Pod". 
Actually maybe the "an Airflow task" is not really needed here? I assume people looking into the KPO know what an Operator is so we could just go with:  

> The KubernetesPodOperator (KPO) runs a Docker image in a dedicated Kubernetes Pod. 

? That way there is no ambiguity. What do you think? 😊

> The [`KubernetesPodOperator`](https://registry.astronomer.io/providers/kubernetes/modules/kubernetesPodoperator) (KPO) is an operator from the Kubernetes provider package designed to provide a straightforward way to execute a task in a Kubernetes Pod from your Airflow environment. In a nutshell, the KPO is an abstraction over a call to the Kubernetes API to launch a Pod.

Reviewer 2: Brilliant! Yeah this makes sense in the context of what operators do - the Bash operator doesn't "Run a task in a bash script", it runs a bash script as a task.

Both the `KubernetesPodOperator` and the `KubernetesExecutor` can use the entire Kubernetes API with regards to Pod creation. Which one to use is an individual choice based on the user's requirements and preferences. You can think of the KPO as "Docker image first" and the `KubernetesExecutor` as "operator first" implementations. The KPO is generally ideal for controlling the environment the task is run in, while the `KubernetesExecutor` is ideal for controlling resource optimization. Frequently the KPO is used in an Airflow environment using the `KubernetesExecutor` to run some tasks where the focus is on environment control while the other tasks are run in Pods via the `KubernetesExecutor`.

## How to configure the `KubernetesPodOperator`

> The KPO launches any valid Docker Image provided to it (passed via the `image` parameter) in a new Kubernetes Pod on an existing Kubernetes cluster. The operator has many other parameters which specify the Pod configuration which the KPO translates into a call to the [Kubernetes API](https://kubernetes.io/docs/reference/kubernetes-api/). The `KubernetesPodOperator` offers full access to Kubernetes API functionality relating to Pod creation.    

Reviewer 2: Ah yeah I forgot about Pod templates! Suggesting something along these lines for my next review

- a `KubeConfig` file, provided as either a path to the file or in JSON format.
- the cluster context to select a specific cluster from the provided `KubeConfig` file.

The picture below shows how to set up a Kubernetes Cluster Connection in the Airflow UI.

![Kubernetes Cluster Connection](https://assets2.astronomer.io/main/guides/kubepod-operator/kubernetes_cluster_connection.png)

The components of the connection can also be set or overwritten at the task level by using the arguments `config_file` (to specify the path to the `KubeConfig` file) and `cluster_context`. Setting these parameters at the level of `airflow.cfg` has been deprecated.

### Example: Using the KPO to Run a Script in Another Language

> A frequent use case for the `KubernetesPodOperator` is to run scripts for tasks better described in a language other than Python. For this purpose a custom Docker image has to be built and either run from a public or private [DockerHub repository](https://docs.docker.com/docker-hub/repos/).

Reviewer 2: I might have asked this question before I got full understanding myself later in my review. I actually think you were quite close the first time around - all you need to do is include the detail that the script is _in_ the Docker image.

However, you may want to use a specific AWS or GCP account that you keep separate from your Airflow environment. Or you are currently running Airflow on a local setup not using Kubernetes, but still want to run some tasks in a Kubernetes Pod on a remote cluster.

The example below shows the steps to take to set up an EKS cluster on AWS and run a Pod on it from an Airflow instance where cross-account access is not feasible. After setup of the EKS cluster, the `KubeConfig` as well as the `AWS Profile` have to be provided to Airflow to make the connection to the remote cluster. Note that the same general steps are applicable to other Kubernetes services (e.g. GKE).  

> **Note**: Some platforms which can host Kubernetes clusters have their own specialised operators available like the [GKEStartPodOperator](https://registry.astronomer.io/providers/google/modules/gkestartPodoperator) and the [EksPodOperator](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/_api/airflow/providers/amazon/aws/operators/eks/index.html#module-airflow.providers.amazon.aws.operators.eks).

### Step 1: Set up an EKS cluster on AWS

> To set up a new EKS cluster on AWS you first need to create an IAM role with suitable permissions.
> On AWS navigate to **Identity and Access Management (IAM)** -> **Roles** -> **Create role**. Select AWS service and `EKS` from the dropdown menu as well as `EKS - Cluster`. Make sure to give the role a unique name and to add the permission policies `AmazonEKSWorkerNodePolicy`, `AmazonEKS_CNI_Policy` and `AmazonEC2ContainerRegistryReadOnly`.

Reviewer 2: ```suggestion
To set up a new EKS cluster on AWS, first [create an EKS cluster IAM role](https://docs.aws.amazon.com/eks/latest/userguide/service_IAM_role.html#create-service-role) with a unique name and to the following permission policies: 

- `AmazonEKSWorkerNodePolicy`
- `AmazonEKS_CNI_Policy`
- `AmazonEC2ContainerRegistryReadOnly`
```

Reviewer 2: I like to link out to documentation for 3rd party setup because you never know when the UI or buttons might change! This helps to future proof docs.

> Edit the trust policy (**IAM** -> **Roles** -> role name) of this new role to include your user and necessary AWS Services. This step ensures that the role can be assumed by your user.

Reviewer 2: ```suggestion
[Update the trust policy](https://docs.aws.amazon.com/directoryservice/latest/admin-guide/edit_trust.html) of this new role to include your user and necessary AWS Services. This step ensures that the role can be assumed by your user.
```

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "AWS": "arn:aws:iam::<aws account id>:<your user>",
        "Service": [
              "ec2.amazonaws.com",
              "eks.amazonaws.com"
        ]
      },
      "Action": "sts:AssumeRole"
    }
  ]
}
```

Add the new role to your local AWS credentials file which by default can be found at `~/.aws/config`.

```text
[default]
region = <your region>

[profile <name of the new role>]
role_arn = <EKS role arn>
source_profile = <your user>
```

> Make sure your default credentials in `~/.aws/credentials` are using a valid and active key for your username (keys can be generated at **IAM** -> **Users** -> your user -> **Security Credentials**).

Reviewer 2: ```suggestion
Make sure your default credentials include a valid and active key for your username (see the [AWS docs](https://docs.aws.amazon.com/kms/latest/developerguide/create-keys.html) for more information).
```

Author: I changed the link here because yours links to the Key Management System docs which I think is a pretty advanced system (I never interacted with it so far). I used the easier way described here: https://docs.aws.amazon.com/general/latest/gr/aws-sec-cred-types.html#access-keys-and-secret-access-keys.

> Make a copy of `~/.aws` available to your Airflow environment (for Astro users: copy it into the `include` folder).

Reviewer 2: ```suggestion
Make a copy of `~/.aws/credentials` available to your Airflow environment. If you run Airflow using the Astro CLI, copy this file into the `include` folder of your Astro project. 
```

Reviewer 2: Is this right?

Reviewer 2: If those are the only two files in the directory then it's probably fine to just add the directory.

Author: They are the only two files I have in that directory but since it is AWS I assume there are many more files you could have in there if you play with other services and settings. So I think specifying the files makes sense :)

> Lastly, create a new EKS cluster (**EKS** -> **Clusters** -> **Add cluster**) and assign the the newly created role to it. The cluster might take a few minutes to become active.

Reviewer 2: ```suggestion
Lastly, [create a new EKS cluster](https://docs.aws.amazon.com/eks/latest/userguide/create-cluster.html) and assign the the newly created role to it. 
```

### Step 2: Retrieve the KubeConfig file from the EKS cluster

> To be able to remotely connect to the newly created EKS cluster the `KubeConfig` file is needed. It can be retrieved by running:

Reviewer 2: ```suggestion
Use a `KubeConfig` file to remotely connect to your new EKS cluster. Retrieve it by running:
```

```bash

> aws eks --region region update-kubeconfig --name cluster_name

Reviewer 2: ```suggestion
aws eks --region <your-region> update-kubeconfig --name <cluster-name>
```

```

> The command above will copy information relating to the new cluster into your existing `KubeConfig` file at `~/.kube/config`. Select the information pertaining to your EKS cluster and make it available to your Airflow environment (for Astro users: copy it into the `include` folder). It is likely that you will have to add the last 5 lines to get to the following structure:

Reviewer 2: ```suggestion
This command copies information relating to the new cluster into your existing `KubeConfig` file at `~/.kube/config`. 

Check this file before making it available to Airflow. It should look like the following configuration. If any of the configurations listed here are missing, add them to the file. 
```

```text
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: <your certificate>
    server: <your AWS server address>
  name: <arn of your cluster>
contexts:
- context:
    cluster: <arn of your cluster>
    user: <arn of your cluster>
  name: <arn of your cluster>
current-context: <arn of your cluster>
kind: Config
preferences: {}
users:
- name: <arn of your cluster>
  user:
    exec:
      apiVersion: client.authentication.k8s.io/v1alpha1
      args:
      - --region
      - <your cluster's AWS region>
      - eks
      - get-token
      - --cluster-name
      - <name of your cluster>
      - --profile
      - default
      command: aws
      interactiveMode: IfAvailable
      provideClusterInfo: false
```

> The `KubeConfig` file is how the KPO will connect to a Kubernetes cluster by running the `awscli` with the default profile, using the credentials you provided in Step 1.

Reviewer 2: ```suggestion
Once you confirm that the `KubeConfig` file is formatted properly, make it available to Airflow. If you're an Astro CLI user, you can do this by adding the file to the `include` directory of your Astro project.
```

### Step 3: Add a namespace and service account to the EKS Cluster

> It is best practice to use a new namespace for all Pods sent to the EKS cluster from the Airflow environment.  

Reviewer 2: ```suggestion
It is best practice to use a new namespace for the Pods that Airflow spins up in your cluster. To create a namespace or your Pods, access the EKS cluster and run:  
```

```bash
# create a new namespace on the EKS cluster

> kubectl create namespace <your namespace name>

Reviewer 2: ```suggestion
kubectl create namespace <your-namespace-name>
```

```

### Step 4: Adjust the Airflow configuration files

> This step will differ depending on the Airflow setup. When running Airflow locally, it is necessary to make sure `awscli`, `apache-airflow-providers-cncf-kubernetes`,
> and `apache-airflow-providers-amazon` are installed correctly on the local machine.

Reviewer 2: ```suggestion
This step will differ depending on the Airflow setup. 

#### Local Apache Airflow 

To access your cluster from local instance of Apache Airflow, install `awscli`, `apache-airflow-providers-cncf-kubernetes`, and `apache-airflow-providers-amazon` on the machine running Airflow.
```

> If you are running Airflow with Docker, add the following line to your Dockerfile to copy your aws credentials from `/include` (or any other location) into the container:

Reviewer 2: ```suggestion
#### Airflow on Docker with the Astro CLI

To access the cluster from Airflow using the Astro CLI, add the following line to your Dockerfile to copy your AWS credentials from into the container:
```
Again, does this have to be `.aws/config`?

```dockerfile
COPY --chown=astro:astro include/.aws /home/astro/.aws
```

> To be able to correctly authenticate yourself to the remote cluster it is also important to add the AWS command line tool (`awscli`) to your `packages.txt`, and the necessary provider packages to `requirements.txt` (the provider packages will differ depending on what cloud you are connecting to):

Reviewer 2: ```suggestion
To authenticate yourself to the remote cluster, add the following line to your `packages.txt` file:
```

```text
awscli
```

```text
apache-airflow-providers-cncf-kubernetes
apache-airflow-providers-amazon
```

### Step 5: Add AWS connection ID

> In the Airflow UI navigate to **Admin** -> **Connections** to set up the connection to the AWS account running the target EKS cluster. Chose a unique connection ID, the Connection Type 'Amazon Web Services' and use your `aws_access_key_id` as login and your `aws_secret_access_key` as password.

Reviewer 2: ```suggestion
In the Airflow UI, go to **Admin** -> **Connections** to set up the connection to the AWS account running the target EKS cluster. Set the following configurations for the connection: 
- **Connection ID**: Any
- **Connection Type**: Amazon Web Services 
- **Login**: Your AWS access key ID
- **Password**: Your AWS secret access key
```

### Step 6: Create the DAG with the KPO

> **Note**: When creating new deployments, users of the Astro cloud will need to set one additional argument in the KPO: `labels={"airflow_kpo_in_cluster": "False"}` to connect to a remote cluster. If you are trying to set up the KPO connecting to a remote cluster in an existing deployment please contact Customer Support.

> If dynamic behavior of the EKS cluster is desired, e.g. Nodes get created with specifications for a task and deleted after the task has run, it is necessary to use several classes from the Amazon provider package as shown in the example DAG below. If your remote Kubernetes cluster has Nodes already available you will only need the `KubernetesPodOperator` itself (task number 3 in the example).

Reviewer 2: ```suggestion
The following DAG utilizes several classes from the Amazon provider package to dynamically spin up and delete Pods for each task in a newly created node group. If your remote Kubernetes cluster already has a node group available, you only need to define your task in the `KubernetesPodOperator` itself.
```

heroImagePath: null
tags: ["Kubernetes", "Operators"]
---

## Overview

> The [KubernetesPodOperator](https://registry.astronomer.io/providers/kubernetes/modules/kubernetesPodoperator) (KPO) runs a Docker image in a Kubernetes Pod as an Airflow task. By abstracting calls to the Kubernetes API, the KubernetesPodOperator enables you to start and run Pods from Airflow using DAG code.

Reviewer 2: ```suggestion
The KubernetesPodOperator (KPO) runs an Airflow task as a Docker image in a dedicated Kubernetes Pod. By abstracting calls to the Kubernetes API, the KubernetesPodOperator enables you to start and run Pods from Airflow using DAG code.
```

Sometimes you may want to run Pods on different clusters, for example if only some need GPU resources while others do not. An example of this is provided at the end of the guide.

### KubernetesPodOperator vs Kubernetes executor

In Airflow you have the option to choose between a variety of [executors](https://www.astronomer.io/guides/airflow-executors-explained) which determine how your Airflow tasks will be executed.

The Kubernetes executor and the KubernetesPodOperator both dynamically launch and terminate Pods to run Airflow tasks. As the name suggests, the Kubernetes executor affects how all tasks in an Airflow instance are executed. The KubernetesPodOperator launches only its own task in a Kubernetes Pod with its own configuration. It does not affect any other tasks in the Airflow instance. Details on how to configure the Kubernetes executor can be found in the [Airflow Documentation](https://airflow.apache.org/docs/apache-airflow/stable/executor/kubernetes.html).

The main differences between the KubernetesPodOperator and the Kubernetes executor are:

> - The Kubernetes executor runs a task defined in Airflow using any existing operator in a Kubernetes Pod without the user needing to specify a Docker image, while the KubernetesPodOperator requires a Docker image to be specified.

Reviewer 2: ```suggestion
- The KubernetesPodOperator requires a Docker image to be specified, while the Kubernetes executor doesn't.
```
Keeping parallel structure with the title. Also cut out some of the details that, while true, don't seem relevant to the main difference.

> - The KubernetesPodOperator defines one isolated Airflow task. In contrast the Kubernetes executor is implemented at the configuration level of the Airflow instance, which means _all_ tasks will each be run in their own Kubernetes Pod. This may be desired in some use cases to leverage auto-scaling, but is generally not ideal for environments with a high volume of shorter running tasks.

Reviewer 2: ```suggestion
- The KubernetesPodOperator defines one isolated Airflow task. In contrast, the Kubernetes executor is implemented at the configuration level of the Airflow instance, which means _all_ tasks run in their own Kubernetes Pod. This might be desired in some use cases to that require auto-scaling, but it's not ideal for environments with a high volume of shorter running tasks.
```

> - In comparison to the KubernetesPodOperator the Kubernetes executor has less abstraction over Pod configuration. All task-level configurations have to be passed in as a dictionary via the `BaseOperator` argument `executor_config`, which is available to all operators.

Reviewer 2: ```suggestion
- In comparison to the KubernetesPodOperator, the Kubernetes executor has less abstraction over Pod configuration. All task-level configurations have to be passed in to the executor as a dictionary using the `BaseOperator's` `executor_config` argument, which is available to all operators.
```

> - If a custom Docker image is passed to the Kubernetes executor's `base` container by providing it to either the `pod_template_file` or the `pod_override` key in the dictionary for the `executor_config` argument, it needs to have Airflow installed, otherwise the task will not run. A possible reason for customizing this Docker image would be to run a task in an environment with different versions of packages than the majority of the tasks an Airflow instance. This is not the case with the KubernetesPodOperator, which can run any valid Docker image.

Reviewer 2: ```suggestion
- If a custom Docker image is passed to the Kubernetes executor's `base` container by providing it to either the `pod_template_file` or the `pod_override` key in the dictionary for the `executor_config` argument, it needs to have Airflow installed, otherwise the task will not run. A possible reason for customizing this Docker image would be to run a task in an environment with different versions of packages than other tasks running in your Airflow instance. This is not the case with the KubernetesPodOperator, which can run any valid Docker image.
```

Both the KubernetesPodOperator and the Kubernetes executor can use the Kubernetes API to create Pods for running tasks. Which one you should use is based on your requirements and preferences. The KubernetesPodOperator is generally ideal for controlling the environment the task is run in, while the Kubernetes executor is ideal for controlling resource optimization. It's common to use both the Kubernetes executor and the KubernetesPodOperator in the same Airflow environment, where all tasks need to run on Kubernetes but only some tasks need the KubernetesPodOperator's fine-grained environment configurations.

## How to configure the KubernetesPodOperator

> The KubernetesPodOperator launches any valid Docker image provided to it in a new Kubernetes Pod on a Kubernetes cluster. The KubernetesPodOperator supports arguments for some of the most common Pod settings, but you can also create a Pod from a YAML file or from Python objects if you need more detailed Pod configuration.

Reviewer 2: ```suggestion
The KubernetesPodOperator launches any valid Docker image provided to it in a dedicated Kubernetes Pod on a Kubernetes cluster. The KubernetesPodOperator supports arguments for some of the most common Pod settings. For more advanced use cases, you can specify a [Pod template file](https://kubernetes.io/docs/concepts/workloads/pods/#pod-templates) that supports all possible Pod settings.
```

The KubernetesPodOperator can be instantiated like any other operator within the context of a DAG. The following are some of the arguments that can be passed to the operator.

### Required arguments

- `task_id`: A unique string identifying the task within Airflow.
- `namespace`: the namespace within your Kubernetes cluster the new Pod should be assigned to.
- `name`: the name of the Pod created. This name needs to be unique for each Pod within a namespace.
- `image`: a Docker image to launch. Images from [hub.docker.com](https://hub.docker.com/) can be passed in with just the image name, custom repositories have to be given as full URLs.

### Optional arguments

- `random_name_suffix`: Generates a random suffix for the Pod name if set to `True`. Avoids naming conflicts when running a large number of Pods.

> - `labels`: Add key:value pairs to the Pod which can be used to logically group decoupled objects together.

Reviewer 2: ```suggestion
- `labels`: A list of key/value pairs which can be used to logically group decoupled objects together.
```

> - `ports`: Provides the possibility to override default ports for the Pod.

Reviewer 2: ```suggestion
- `ports`: Ports for the Pod.
```
When defining configurations, it's always best to strike at the exact thing the user has to define. If the user has to specify an object, the definition of the config should focus on that object.

> - `reattach_on_restart`: Defines how to handle losing the worker while the Pod is running, by default (`True`) the existing Pod will reattach to the worker on the next try. `False` creates a new Pod for each try.

Reviewer 2: ```suggestion
- `reattach_on_restart`: Defines how to handle losing the worker while the Pod is running.  When set to `True`, the existing Pod reattaches to the worker on the next try. When set to `False`, a new Pod will be created for each try. The default is `True`.
```

> - `is_delete_operator_pod`: `True` by default, will delete the Pod when it reaches its final state or when the execution is interrupted.

Reviewer 2: ```suggestion
- `is_delete_operator_pod`: Determines whether to delete the Pod when it reaches its final state or when the execution is interrupted. The default is `True`.
```

> - `get_logs`: Provides the `stdout` of the container as task-logs to the Airflow logging system.

Reviewer 2: ```suggestion
- `get_logs`: Determines whether to use the `stdout` of the container as task-logs to the Airflow logging system.
```

> - `log_events_on_failure`: if set to `True` events will be logged in case the Pod fails (default: `False`).

Reviewer 2: ```suggestion
- `log_events_on_failure`: Determines whether are logged in case the Pod fails. The default is `False`.
```

> - `env_vars`: Allows the user to specify individual environment variables for a Pod.

Reviewer 2: ```suggestion
- `env_vars`: A list of environment variables for the Pod.
```

> - `resources`: Allows the user to pass a dictionary with resource requests (keys: `request_memory`, `request_cpu`) and limits (keys: `limit_memory`, `limit_cpu`, `limit_gpu`). See the [Kubernetes Documentation on Resource Management for Pods and Containers](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/) for more information.

Reviewer 2: ```suggestion
- `resources`: A dictionary with resource requests (keys: `request_memory`, `request_cpu`) and limits (keys: `limit_memory`, `limit_cpu`, `limit_gpu`). See the [Kubernetes Documentation on Resource Management for Pods and Containers](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/) for more information.
```

> - `volumes`: Can be used to pass in a modified `k8s.V1Volume`, see also the [Kubernetes example DAG from the Airflow documentation](https://airflow.apache.org/docs/apache-airflow-providers-cncf-kubernetes/stable/_modules/tests/system/providers/cncf/kubernetes/example_kubernetes.html).

Reviewer 2: ```suggestion
- `volumes`: A list of `k8s.V1Volumes`, see also the [Kubernetes example DAG from the Airflow documentation](https://airflow.apache.org/docs/apache-airflow-providers-cncf-kubernetes/stable/_modules/tests/system/providers/cncf/kubernetes/example_kubernetes.html).
```

> - `affinity` and `tolerations` are ways to further specify rules for [Pod to Node assignment](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/). Like the `volumes` parameter they also require their input as a `k8s` object.

Reviewer 2: ```suggestion
- `affinity` and `tolerations`: Dictionaries of rules for [Pod to Node assignments](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/). Like the `volumes` parameter, these also require a `k8s` object.
```

> - `pod_template_file`: Provides a way to specify the path to a YAML file defining Pod configurations.

Reviewer 2: ```suggestion
- `pod_template_file`: The path to a Pod template file.
```

> - `full_pod_spec`: Allows the user to specify Pod configurations using Python `k8s` objects.

Reviewer 2: ```suggestion
- `full_pod_spec`: A complete Pod configuration formatted as a Python `k8s` object.
```

- A `KubeConfig` file, provided as either a path to the file or in JSON format.
- The cluster context from the provided `KubeConfig` file.

The following image shows how to set up a Kubernetes cluster connection in the Airflow UI.

![Kubernetes Cluster Connection](https://assets2.astronomer.io/main/guides/kubepod-operator/kubernetes_cluster_connection.png)

The components of the connection can also be set or overwritten at the task level by using the arguments `config_file` (to specify the path to the `KubeConfig` file) and `cluster_context`. Setting these parameters at the level of `airflow.cfg` has been deprecated.

### Example: Using the KubernetesPodOperator to Run a Script in Another Language

> A frequent use case for the KubernetesPodOperator is to run a Docker image containing a script for a task better described in a language other than Python in a Kubernetes Pod. To do this, you build a custom Docker image containing one or more scripts in your preferred language plus instructions on how to run these scripts and run it from either a public or private [DockerHub repository](https://docs.docker.com/docker-hub/repos/).

Reviewer 2: ```suggestion
A frequent use case for the KubernetesPodOperator is run a task in a language other than Python. To do this, you build a custom Docker image containing the script.
```

CMD ["python", "./multiply_by_23.py"]
```

When using XComs with the KubernetesPodOperator, you must create the file `airflow/xcom/return.json` in your Docker container (ideally from within your Dockerfile as seen above), because Airflow can only look for XComs to pull at that specific location. The Docker image also contains a simple Python script to multiply an environment variable by 23, package the result into JSON, and write that JSON to the correct file to be retrieved as an XCom. The XComs from the KubernetesPodOperator are pushed only if the task itself was marked as successful.  

```python
import os

# import the result of the previous task as an environment variable
data_point = os.environ['DATA_POINT']
multiplied_data_point = 23 * int(data_point)

# multiply the data point by 23 and package the result into a json
multiplied_data_point = str(23 * int(data_point))
return_json = {"return_value":f"{multiplied_data_point}"}

# write to the file checked by airflow for XComs
f = open('./airflow/xcom/return.json', 'w')
f.write(f"{return_json}")
f.close()
```

Lastly the `load_data` task pulls the XCom returned from the `transform` task and prints it to the console.

> Below you can find the full DAG code. Remember to only turn on `do_xcom_push` if you have created the `airflow/xcom/return.json` within the Docker container run by the KubernetesPodOperator, otherwise the task will fail.

Reviewer 2: ```suggestion
Below you can find the full DAG code. Remember to turn on `do_xcom_push` only if you have created the `airflow/xcom/return.json` within the Docker container run by the KubernetesPodOperator. Otherwise, the task will fail.
```

```python
from airflow import DAG
from datetime import datetime
from airflow.providers.cncf.kubernetes.operators.kubernetes_pod import KubernetesPodOperator
from airflow.configuration import conf
from airflow.decorators import task

import random

# get the current Kubernetes namespace Airflow is running in
namespace = conf.get("kubernetes", "NAMESPACE")

> # instatiate the DAG

Reviewer 2: ```suggestion
# instantiate the DAG
```

with DAG(
    start_date=datetime(2022,6,1),
    catchup=False,
    schedule_interval='@daily',
    dag_id='KPO_example_DAG'
) as dag:

    @task
    def extract_data():
        # simulating querying from a database
        data_point = random.randint(0,100)
        return data_point

    transform = KubernetesPodOperator(
        # operator specific argument
        task_id='transform',

        # arguments pertaining to the image and commands executed
        image='< image location >', # the Docker Image to launch

        # arguments pertaining to where the Pod is launched
        in_cluster=True, # launch the Pod on the same cluster as Airflow is running on
        namespace=namespace, # launch the Pod in the same namespace as Airflow is running in

        # Pod configuration
        name='my_pod', # naming the Pod
        get_logs=True, # log stdout of the container as task logs
        log_events_on_failure=True, #log events in case of Pod failure
        env_vars={"DATA_POINT": "{{ ti.xcom_pull(task_ids='extract_data', key='return_value') }}"},

>         do_xcom_push=True #push the contents from xcom.json to xcoms

Reviewer 2: Worth repeating the note above in the comment here?

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "AWS": "arn:aws:iam::<aws account id>:<your user>",
        "Service": [
              "ec2.amazonaws.com",
              "eks.amazonaws.com"
        ]
      },
      "Action": "sts:AssumeRole"
    }
  ]
}
```

Add the new role to your local AWS config file, which by default is located at `~/.aws/config` (for more information see the [AWS docs](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html#cli-configure-files-where)).

```text
[default]
region = <your region>

[profile <name of the new role>]
role_arn = <EKS role arn>
source_profile = <your user>
```

Make sure your default credentials include a valid and active key for your username (see the [AWS docs](https://docs.aws.amazon.com/general/latest/gr/aws-sec-cred-types.html#access-keys-and-secret-access-keys) for more information).

Make a copy of `~/.aws/credentials` and `~/.aws/config` available to your Airflow environment. If you run Airflow using the Astro CLI, create a new directory called `.aws` in the `include` folder of your Astro project and copy both files into it.

> Lastly, [create a new EKS cluster](https://docs.aws.amazon.com/eks/latest/userguide/create-cluster.html) and assign the the newly created role to it.

Reviewer 2: ```suggestion
Lastly, [create a new EKS cluster](https://docs.aws.amazon.com/eks/latest/userguide/create-cluster.html) and assign the newly created role to it.
```

    user: <arn of your cluster>
  name: <arn of your cluster>
current-context: <arn of your cluster>
kind: Config
preferences: {}
users:
- name: <arn of your cluster>
  user:
    exec:
      apiVersion: client.authentication.k8s.io/v1alpha1
      args:
      - --region
      - <your cluster's AWS region>
      - eks
      - get-token
      - --cluster-name
      - <name of your cluster>
      - --profile
      - default
      command: aws
      interactiveMode: IfAvailable
      provideClusterInfo: false
```

Once you confirm that the `KubeConfig` file is formatted properly, make it available to Airflow. If you're an Astro CLI user, you can do this by adding the file to the `include` directory of your Astro project.

### Step 3: Add a new namespace to the EKS Cluster

It is best practice to use a new namespace for the Pods that Airflow spins up in your cluster. To create a namespace for your Pods, access the EKS cluster and run:  

```bash
# create a new namespace on the EKS cluster
kubectl create namespace <your-namespace-name>
```

### Step 4: Adjust the Airflow configuration files

This step will differ depending on the Airflow setup.

#### Local Apache Airflow

> To access your cluster from local instance of Apache Airflow, install `awscli`, `apache-airflow-providers-cncf-kubernetes`, and `apache-airflow-providers-amazon` on the machine running Airflow.

Reviewer 2: ```suggestion
To access your cluster from a local instance of Apache Airflow, install `awscli`, `apache-airflow-providers-cncf-kubernetes`, and `apache-airflow-providers-amazon` on the machine running Airflow.
```
